### PR TITLE
feat: build apisix-openresty-debug

### PR DESCRIPTION
**Pull request by tzssangglass**
add build apisix-openresty-debug and apisix-openresty-asan shell

### CODE REVIEW

File: build-apisix-openresty-asan-centos7.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+set -euo pipefail
+set -x
+
+yum-config-manager --add-repo https://openresty.org/package/centos/openresty.repo
+yum install centos-release-scl -y
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:6]: Why do we need to install this?

File: build-apisix-openresty-asan-centos7.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+export openssl_prefix=/usr/local/openresty/openssl111
+export zlib_prefix=/usr/local/openresty/zlib
+export pcre_prefix=/usr/local/openresty/pcre
+
+export ASAN_OPTIONS=detect_leaks=0
+export OR_PREFIX="/usr/local/openresty-asan"
+export cc_opt="-DNGX_LUA_ABORT_AT_PANIC -I${zlib_prefix}/include -I${pcre_prefix}/include -I${openssl_prefix}/include -O1"
+export ld_opt="-L${zlib_prefix}/lib -L${pcre_prefix}/lib -L${openssl_prefix}/lib -Wl,-rpath,${zlib_prefix}/lib:${pcre_prefix}/lib:${openssl_prefix}/lib"
+export cc='--with-cc="ccache gcc -fsanitize=address"'
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:18]: The official openresty-asan uses clang to build itself

File: build-apisix-openresty-asan-centos7.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+export luajit_xcflags="-DLUAJIT_NUMMODE=2 -DLUAJIT_ENABLE_LUA52COMPAT -DLUAJIT_USE_VALGRIND -O1 -fno-omit-frame-pointer"
+export nproc="-j`nproc`"
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:20]: Let's move the common nproc argument to the build-apisix-openresty.sh

File: build-apisix-openresty-asan-centos7.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+export no_pool_patch="--with-no-pool-patch"
+
+./build-apisix-openresty.sh latest
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:23]: We can't use latest which don't take care about the tag

Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:23]: We can't use the latest which don't take care of the tag

File: build-apisix-openresty-asan-centos7.sh (hunk 3 above)
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:20]: Let's move the common nproc argument to the build-apisix-openresty.sh

File: build-apisix-openresty-debug-centos7.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+set -euo pipefail
+set -x
+
+yum-config-manager --add-repo https://openresty.org/package/centos/openresty.repo
+yum -y install gcc gcc-c++ patch wget git make sudo
+yum -y install openresty-openssl111-devel openresty-pcre-devel openresty-zlib-devel
```
Comment (spacewander) [inline build-apisix-openresty-debug-centos7.sh:7]: Should be openresty-openssl111-debug-devel

File: build-apisix-openresty-asan-centos7.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+set -euo pipefail
+set -x
+
+yum-config-manager --add-repo https://openresty.org/package/centos/openresty.repo
+yum install centos-release-scl -y
+yum -y install gcc gcc-c++ patch wget git make sudo libasan
+yum -y install openresty-openssl111-asan openresty-pcre-asan openresty-zlib-asan
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:8]: The dependencies should have -devel suffix.

File: build-apisix-openresty-asan-centos7.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+export openssl_prefix=/usr/local/openresty/openssl111
```
Comment (spacewander) [inline build-apisix-openresty-asan-centos7.sh:10]: The prefix should be openresty-asan/openssl111

File: build-apisix-openresty-debug-centos7.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+export openssl_prefix=/usr/local/openresty/openssl111
```
Comment (spacewander) [inline build-apisix-openresty-debug-centos7.sh:9]: Ditto

**Review by spacewander — CHANGES_REQUESTED** [review]

File: build-apisix-openresty-asan-centos7.sh (hunk 2 above)
Comment (tzssangglass) [reply build-apisix-openresty-asan-centos7.sh:18]: I saw this https://github.com/openresty/openresty-packaging/commit/d7fb0a7814151033aa0c5babbb9fd40e48f49905. 
and I'm building `openresty-asan` on an arm64 machine and it's confusing me. It always prompts me with
 `Error: Package: openresty-openssl111-asan-1.1.1k-1.el7.aarch64 (openresty)
           Requires: libasan
`
later I will try to build `openresty-asan` on an x86_64 machine

File: build-apisix-openresty-asan-centos7.sh (hunk 7 above)
Comment (tzssangglass) [reply build-apisix-openresty-asan-centos7.sh]: update

File: build-apisix-openresty-asan-centos7.sh (hunk 3 above)
Comment (tzssangglass) [reply build-apisix-openresty-asan-centos7.sh]: update

File: build-apisix-openresty-asan-centos7.sh (hunk 4 above)
Comment (tzssangglass) [reply build-apisix-openresty-asan-centos7.sh]: update

File: build-apisix-openresty-asan-centos7.sh (hunk 1 above)
Comment (tzssangglass) [reply build-apisix-openresty-asan-centos7.sh]: update

**Comment by tzssangglass** [comment]
@spacewander , I fount it hard to build `openresty-asan` for me, and I would like to keep only the code to build `openresty-debug` in this PR, can you propose a separate PR to build `openresty-asan`.

File: build-apisix-openresty-debug-centos7.sh [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+export cc_opt="-DNGX_LUA_USE_ASSERT -DNGX_LUA_ABORT_AT_PANIC -I${zlib_prefix}/include -I${pcre_prefix}/include -I${openssl_prefix}/include -O0"
+export ld_opt="-L${zlib_prefix}/lib -L${pcre_prefix}/lib -L${openssl_prefix}/lib -Wl,-rpath,${zlib_prefix}/lib:${pcre_prefix}/lib:${openssl_prefix}/lib"
+export luajit_xcflags="-DLUAJIT_ASSERT -DLUAJIT_NUMMODE=2 -DLUAJIT_ENABLE_LUA52COMPAT -O0"
+export OR_PREFIX=/usr/local/openresty-debug
+export debug_args=--with-debug
+
+./build-apisix-openresty.sh
+
+ln -sf /usr/local/openresty-debug/nginx/sbin/nginx /usr/bin/openresty-debug
```
Comment (spacewander) [inline build-apisix-openresty-debug-centos7.sh:21]: the binary is still /usr/bin/openresty. BTW, this job should be done by the post-install script.

Comment (tzssangglass) [reply build-apisix-openresty-debug-centos7.sh:21]: that's right, then I'll delete this

**Review by spacewander — APPROVED** [review]

**Pull request merged**